baseplate/experiments/targeting/tree_targeting.py

```diff
@@ -0,0 +1,192 @@
+                return True
+
+        return False
+
+
+class NotNode(Targeting):
+    """Boolean 'not' operator"""
+    def __init__(self, input_node):
+        """Build a node that evaluates to true if its sole child node evaluates
+        to False (boolean 'not')
+
+        :param dict input_node: a Targeting node
+        """
+        if len(input_node) != 1:
+            raise ValueError("NotNode expects exactly one field.")
+
+        if not isinstance(input_node, dict):
+            raise TypeError("Input to NotNode expects a dictionary")
+
+        self._child = create_targeting_tree(input_node)
+
+    def evaluate(self, **kwargs):
+        return not self._child.evaluate(**kwargs)
+
+
+class OverrideNode(Targeting):
+    """Always return True/False
+    """
+    def __init__(self, input_node):
+        if input_node is True:
+            self._return_value = True
+        else:
+            self._return_value = False
+
+    def evaluate(self, **kwargs):
+        return self._return_value
+
+
+OPERATOR_NODE_TYPE_MAPPING = {
+    "ANY": AnyNode,
+    "ALL": AllNode,
+    "EQ": EqualNode,
+    "NOT": NotNode,
+    "OVERRIDE": OverrideNode,
+}
+
+
+def create_targeting_tree(input_node):
+    """Creates a tree-based targeting evaluator.
+
+    Processes input json to create a tree against which a set of inputs
+    can be evaluated to determine whether or not a user with the input
+    input attributes should be targeted for an experiment.
+
+    Each node is represented by a dict with one key which represents the
+    operator. The value for the operator is either a dictionary (another
+    node) or a list of dictionaries (each of which is another node).
+
+    An example of a targeting tree config might look like:
+
+    targeting_cfg = {
+        'ALL':[
+            {'ANY':[
+                {'EQ': {'field': 'is_mod', 'value': True}},
+                {'EQ': {'field': 'user_id', 'values':['t2_1','t2_2','t2_3','t2_4']}},
+            ]},
+            {'NOT': {
+                'EQ': {'field': 'has_commented', 'value': True}}},
+            {'EQ': {'field': 'is_logged_in', 'values': [True, False]}},
+            {'NOT': {
+                'EQ': {'field': 'subreddit_id', 'values': ['t5_1','t5_2']}}},
+            {'ALL':[
+                {'EQ':{'field':'votes_cast', 'values':[1,2,3,4,5]}},
+                {'EQ':{'field':'votes_cast', 'value':5}}
+            ]}
+        ]
+    }
+
+    In the above example, we see that all of the following conditions must be
+    met in order for a user to be targeted:
+
+    1. a user must be a mod or have the user id between 1 and 4
+    2. the must not have commented
+    3. they can be logged in or logged out
+    4. they must have cast 5 votes
+
+    """
+    if not len(input_node) == 1:
+        raise ValueError("Call to create_targeting_tree expects a single input key.")
+
+    operator = list(input_node.keys())[0]
+    input_node_value = input_node[operator]
```

Comment:
You could do this in one line as `operator, input_node_value = list(input_node.values())[0]`.

Reply:
Doesn't `list(input_node.values())[0]` just return the first _value_? I need both the key and value of the first (only) item in the dict.

Reply:
Oops, yes that is exactly what I meant.